File: frontend/src/services/genkit.ts
import { auth } from '../config/firebase';

// Configuration
const API_URL = import.meta.env.VITE_API_URL || 'http://localhost:8080/api';
const USE_MOCK = import.meta.env.VITE_USE_MOCK_API !== 'false';

// Helper for Auth Token
const getAuthToken = async () => {
  if (import.meta.env.DEV) return 'dev-token';
  const user = auth.currentUser;
  return user ? await user.getIdToken() : '';
};

// Types
export interface CoverLetterRequest {
  candidate_profile: any;
  job_description: string;
  company_info?: any;
  style?: string;
  format_type?: string;
  special_instructions?: string;
}

export interface KSCRequest {
  user_profile_data: any;
  ksc_statement: string;
}

export interface SmartCoverLetter {
  letter_content: string;
  subject_line?: string;
  sections: any[];
  analysis: any;
  personalization_notes: string[];
  key_selling_points: string[];
  company_connections: string[];
  alternative_versions: Record<string, string>;
  follow_up_suggestions: string[];
}

export interface STAR_Response {
  situation: string;
  task: string;
  action: string;
  result: string;
}

export interface OptimizeResumeRequest {
  resume_text: string;
  missing_keywords: string[];
  job_description: string;
}

export interface OptimizedResume {
  resume_text: string;
  keywords_integrated: string[];
}

export interface CompanyContextRequest {
  company_name: string;
  job_description: string;
}

export interface CompanyContext {
  recent_achievements: string[];
  core_values: string[];
  recommended_tone: string;
  why_work_here_points: string[];
  interview_questions: string[];
  cultural_insights: string;
}

export interface JobListingDetails {
  due_date?: string;
  company_name?: string;
  role_title?: string;
  hiring_manager?: string;
  manager_contact?: string;
  essential_criteria: string[];
  desirable_criteria: string[];
  role_type?: string;
  subsectors: string[];
  location?: string;
  key_responsibilities: string[];
  full_description?: string;
}

export interface UnifiedJobAnalysis {
  job_details: JobListingDetails;
  company_context?: CompanyContext;
  analysis_success: boolean;
  error_message?: string;
}

export interface AnalyzeJobUrlRequest {
  url: string;
}

// Mock Implementation
const mockGenkitApi = {
  async generateCoverLetter(data: CoverLetterRequest): Promise<SmartCoverLetter> {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    return {
      letter_content: `Dear Hiring Manager,\n\nI am writing to express my interest in the position... (Mock Generated for ${data.style})`,
      sections: [],
      analysis: {
        readability_score: 85,
        personalization_score: 90,
        compelling_score: 88,
        keyword_alignment: 92,
        strengths: ['Strong opening', 'Clear examples'],
        improvement_areas: [],
        tone_assessment: 'Professional',
        unique_elements: [],
      },
      personalization_notes: ['Matched skills to job desc'],
      key_selling_points: ['Leadership', 'Technical Skills'],
      company_connections: ['Values alignment'],
      alternative_versions: {},
      follow_up_suggestions: ['Send thank you email'],
    };
  },

  async generateKSCResponse(criteria: string, starData: any): Promise<STAR_Response> {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    return {
      situation: starData.situation || 'Mock Situation',
      task: starData.task || 'Mock Task',
      action: starData.action || 'Mock Action',
      result: starData.result || 'Mock Result - Achieved 100% efficiency',
    };
  },

  async optimizeResume(data: OptimizeResumeRequest): Promise<OptimizedResume> {
    await new Promise((resolve) => setTimeout(resolve, 2000));
    return {
      resume_text: `${data.resume_text}\n\n[Mock: Integrated keywords: ${data.missing_keywords.join(', ')}]`,
      keywords_integrated: data.missing_keywords,
    };
  },

  async getCompanyContext(data: CompanyContextRequest): Promise<CompanyContext> {
    await new Promise((resolve) => setTimeout(resolve, 1500));
    return {
      recent_achievements: [
        `${data.company_name} launched new product line in Q3 2025`,
        'Expanded to APAC market',
      ],
      core_values: ['Innovation', 'Collaboration', 'Customer-First'],
      recommended_tone: 'conversational',
      why_work_here_points: [
        'Leading innovator in the space',
        'Strong growth trajectory',
        'Excellent learning opportunities',
      ],
      interview_questions: [
        'How does the team approach innovation?',
        'What are the biggest challenges for this role?',
        'How do you measure success in this position?',
      ],
      cultural_insights: 'Fast-paced, collaborative environment with emphasis on autonomy',
    };
  },

  async analyzeJobFromUrl(data: AnalyzeJobUrlRequest): Promise<UnifiedJobAnalysis> {
    await new Promise((resolve) => setTimeout(resolve, 3000));
    return {
      job_details: {
        company_name: 'Example Corp',
        role_title: 'Senior Software Engineer',
        essential_criteria: ['Python', 'React', '5+ years experience'],
        desirable_criteria: ['AWS', 'Docker', 'Team leadership'],
        role_type: 'Full-time',
        subsectors: ['Technology', 'SaaS'],
        location: 'Remote',
        key_responsibilities: [
          'Lead development of core platform features',
          'Mentor junior engineers',
          'Collaborate with product team',
        ],
        full_description: `[Mock] Full job description extracted from ${data.url}`,
      },
      company_context: {
        recent_achievements: [
          'Launched AI-powered analytics platform in Q4 2025',
          'Expanded to 15 new markets',
        ],
        core_values: ['Innovation', 'Customer Success', 'Collaboration'],
        recommended_tone: 'conversational',
        why_work_here_points: [
          'Cutting-edge technology stack',
          'Strong growth trajectory',
          'Remote-first culture',
        ],
        interview_questions: [
          'What excites you most about our product roadmap?',
          'How do you approach technical mentorship?',
          'What are your thoughts on our tech stack?',
        ],
        cultural_insights: 'Fast-paced startup environment with strong emphasis on innovation',
      },
      analysis_success: true,
    };
  },
};

// Real Implementation
const realGenkitApi = {
  async generateCoverLetter(data: CoverLetterRequest): Promise<SmartCoverLetter> {
    const token = await getAuthToken();
    const response = await fetch(`${API_URL}/genkit/cover-letter/generate`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || 'Failed to generate cover letter');
    }

    return await response.json();
  },

  async generateKSCResponse(
    criteria: string,
    starData: any,
    userProfile: any
  ): Promise<STAR_Response> {
    const token = await getAuthToken();
    const payload: KSCRequest = {
      user_profile_data: userProfile,
      ksc_statement: `Criteria: ${criteria}\nSituation: ${starData.situation}\nTask: ${starData.task}\nAction: ${starData.action}\nResult: ${starData.result}`,
    };

    const response = await fetch(`${API_URL}/genkit/ksc/generate`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || 'Failed to generate KSC response');
    }

    return await response.json();
  },

  async optimizeResume(data: OptimizeResumeRequest): Promise<OptimizedResume> {
    const token = await getAuthToken();
    const response = await fetch(`${API_URL}/genkit/resume/optimize`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || 'Failed to optimize resume');
    }

    return await response.json();
  },

  async getCompanyContext(data: CompanyContextRequest): Promise<CompanyContext> {
    const token = await getAuthToken();
    const response = await fetch(`${API_URL}/genkit/company/context`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || 'Failed to get company context');
    }

    return await response.json();
  },

  async analyzeJobFromUrl(data: AnalyzeJobUrlRequest): Promise<UnifiedJobAnalysis> {
    const token = await getAuthToken();
    const response = await fetch(`${API_URL}/genkit/job/analyze-url`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({}));
      throw new Error(errorData.detail || 'Failed to analyze job from URL');
    }

    return await response.json();
  },
};

export const genkitApi = USE_MOCK ? mockGenkitApi : realGenkitApi;

File: backend/app/api/endpoints/genkit.py
"""
API Endpoints for Genkit AI Flows
"""

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel

from app.api.endpoints._shared import run_genkit_endpoint
from app.core.genkit_init import is_genkit_enabled
from app.genkit_flows.company_context import (
    CompanyContext,
    generate_company_context,
)
from app.genkit_flows.ksc_generator import (
    STAR_Response,
    generateKscResponse,
)
from app.genkit_flows.resume_optimizer import (
    OptimizedResume,
    optimize_resume,
)
from app.genkit_flows.smart_cover_letter_system import (
    SmartCoverLetter,
    generate_smart_cover_letter,
)
from app.genkit_flows.unified_job_analyzer import (
    UnifiedJobAnalysis,
    analyze_job_from_url,
)

router = APIRouter()


class CoverLetterRequest(BaseModel):
    candidate_profile: dict[str, Any]
    job_description: str
    company_info: dict[str, Any] | None = None
    style: str = "professional"
    format_type: str = "full_letter"
    special_instructions: str | None = None


class KSCRequest(BaseModel):
    user_profile_data: dict[str, Any]
    ksc_statement: str


@router.post(
    "/cover-letter/generate",
    response_model=SmartCoverLetter,
    summary="Generate Smart Cover Letter",
    description="Generates a personalized cover letter using Genkit AI.",
)
async def generate_cover_letter_endpoint(request: CoverLetterRequest) -> SmartCoverLetter:
    async def operation() -> SmartCoverLetter:
        # Note: generate_smart_cover_letter is currently synchronous.
        return generate_smart_cover_letter(
            candidate_profile=request.candidate_profile,
            job_description=request.job_description,
            company_info=request.company_info,
            style=request.style,
            format_type=request.format_type,
            special_instructions=request.special_instructions,
        )

    return await run_genkit_endpoint(
        operation,
        "Cover letter generation failed",
        enabled_check=is_genkit_enabled,
    )


@router.post(
    "/ksc/generate",
    response_model=STAR_Response,
    summary="Generate KSC Response",
    description="Generates a STAR-formatted KSC response using Genkit AI.",
)
async def generate_ksc_endpoint(request: KSCRequest) -> STAR_Response:
    async def operation() -> STAR_Response:
        return await generateKscResponse(
            user_profile_data=request.user_profile_data,
            ksc_statement=request.ksc_statement,
        )

    return await run_genkit_endpoint(
        operation,
        "KSC generation failed",
        enabled_check=is_genkit_enabled,
    )


class OptimizeResumeRequest(BaseModel):
    resume_text: str
    missing_keywords: list[str]
    job_description: str


class CompanyContextRequest(BaseModel):
    company_name: str
    job_description: str


class AnalyzeJobUrlRequest(BaseModel):
    url: str


@router.post(
    "/job/analyze-url",
    response_model=UnifiedJobAnalysis,
    summary="Analyze Job from URL",
    description="One-stop analysis: scrapes URL, extracts job details, and generates company context.",
)
async def analyze_job_url_endpoint(request: AnalyzeJobUrlRequest) -> UnifiedJobAnalysis:
    async def operation() -> UnifiedJobAnalysis:
        return await analyze_job_from_url(url=request.url)

    return await run_genkit_endpoint(
        operation,
        "Job URL analysis failed",
        enabled_check=is_genkit_enabled,
    )


@router.post(
    "/resume/optimize",
    response_model=OptimizedResume,
    summary="Optimize Resume",
    description="Optimizes resume by integrating missing keywords from ATS analysis.",
)
async def optimize_resume_endpoint(request: OptimizeResumeRequest) -> OptimizedResume:
    async def operation() -> OptimizedResume:
        return await optimize_resume(
            resume_text=request.resume_text,
            missing_keywords=request.missing_keywords,
            job_description=request.job_description,
        )

    return await run_genkit_endpoint(
        operation,
        "Resume optimization failed",
        enabled_check=is_genkit_enabled,
    )


@router.post(
    "/company/context",
    response_model=CompanyContext,
    summary="Get Company Context",
    description="Generates company context for applications and interview prep using AI knowledge.",
)
async def get_company_context_endpoint(request: CompanyContextRequest) -> CompanyContext:
    async def operation() -> CompanyContext:
        return await generate_company_context(
            company_name=request.company_name,
            job_description=request.job_description,
        )

    return await run_genkit_endpoint(
        operation,
        "Company context generation failed",
        enabled_check=is_genkit_enabled,
    )
